Unhelpful message with a very long tuple initializer

This program genenerates an unhelpful error message:
```py
from typing import Tuple
t: Tuple[int, ...] = (
    1, 2, 3, 4, 5, 6, 7, 8, 9., 10, 11, 12, 13, 14, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
    30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
    3, 31, 32, 33, 34, 35, 36, 37, 3, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
    30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 4, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
    30, 31, 32, 33, 34, 234234, 36, 37, 38, 39, 40, 1, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56,
    57, 58, 59, 60, 61, 62, 63
)
```

This is the message:

```
t.py:3: error: Incompatible types in assignment (expression has type <tuple: 148 items>, variable has type "Tuple[int, ...]")
```

This is derived from a real example, where the type was `Tuple[Type[x], ...]` and the initializer contained a long tuple of type objects. Figuring out which type object is incompatible was tricky.

Some ideas about how to improve this:
* Point to individual tuple items that are incompatible. For example: `Tuple item 8 has type "float"; "int" expected` (also point the error message to the relevant line/column).
* If there are many incompatibilities, only show a fixed number of them and hide the remaining ones.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4027,7 +4027,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
         subtype = get_proper_type(subtype)
         supertype = get_proper_type(supertype)
-        if self.check_long_tuple_assignment(subtype, supertype, context, msg,
+        if self.msg.try_report_long_tuple_assignment_error(subtype, supertype, context, msg,
                                        subtype_label, supertype_label, code=code):
             return False
         if self.should_suppress_optional_error([subtype]):
@@ -4379,73 +4379,6 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         yes_map, no_map = conditional_type_map(expr, vartype, type)
         yes_map, no_map = map(convert_to_typetype, (yes_map, no_map))
         return yes_map, no_map
-
-    def check_long_tuple_assignment(self,
-                                    subtype: ProperType,
-                                    supertype: ProperType,
-                                    context: Context,
-                                    msg: str = message_registry.INCOMPATIBLE_TYPES,
-                                    subtype_label: Optional[str] = None,
-                                    supertype_label: Optional[str] = None,
-                                    code: Optional[ErrorCode] = None) -> bool:
-        """Check if assignment error using long tuple initializers."""
-        if isinstance(subtype, TupleType):
-            if (len(subtype.items) > 10 and
-                isinstance(supertype, Instance) and
-                    supertype.type.fullname == 'builtins.tuple'):
-                lhs_type = supertype.args[0]
-                lhs_types = [lhs_type] * len(subtype.items)
-                self.generate_incompatible_tuple_error(lhs_types,
-                                    subtype.items, context, msg, code)
-                return True
-            elif (isinstance(supertype, TupleType) and
-                    (len(subtype.items) > 10 or len(supertype.items) > 10)):
-                if len(subtype.items) != len(supertype.items):
-                    if supertype_label is not None and subtype_label is not None:
-                        error_msg = "{} ({} {}, {} {})".format(msg, subtype_label,
-                                        self.format_long_tuple_type(subtype), supertype_label,
-                                        self.format_long_tuple_type(supertype))
-                        self.fail(error_msg, context, code=code)
-                        return True
-                self.generate_incompatible_tuple_error(supertype.items,
-                                    subtype.items, context, msg, code)
-                return True
-        return False
-
-    def format_long_tuple_type(self, typ: TupleType) -> str:
-        """Format very long tuple type using an ellipsis notation"""
-        item_cnt = len(typ.items)
-        if item_cnt > 10:
-            return 'Tuple[{}, {}, ... <{} more items>]'\
-                    .format(format_type_bare(typ.items[0]),
-                        format_type_bare(typ.items[1]), str(item_cnt - 2))
-        else:
-            return format_type_bare(typ)
-
-    def generate_incompatible_tuple_error(self,
-                                          lhs_types: List[Type],
-                                          rhs_types: List[Type],
-                                          context: Context,
-                                          msg: str = message_registry.INCOMPATIBLE_TYPES,
-                                          code: Optional[ErrorCode] = None) -> None:
-        """Generate error message for individual incompatible tuple pairs"""
-        error_cnt = 0
-        notes = []  # List[str]
-        for i, (lhs_t, rhs_t) in enumerate(zip(lhs_types, rhs_types)):
-            if not is_subtype(lhs_t, rhs_t):
-                if error_cnt < 3:
-                    notes.append('Expression tuple item {} has type "{}"; "{}" expected; '
-                        .format(str(i), format_type_bare(rhs_t), format_type_bare(lhs_t)))
-                error_cnt += 1
-
-        error_msg = msg + ' ({} tuple items are incompatible'.format(str(error_cnt))
-        if error_cnt - 3 > 0:
-            error_msg += '; {} items are omitted)'.format(str(error_cnt - 3))
-        else:
-            error_msg += ')'
-        self.fail(error_msg, context, code=code)
-        for note in notes:
-            self.note(note, context, code=code)
 
 
 def conditional_type_map(expr: Expression,

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4388,6 +4388,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                                subtype_label: Optional[str] = None,
                                supertype_label: Optional[str] = None,
                                code: Optional[ErrorCode] = None) -> bool:
+        """Check if assignment error using long tuple initializers."""
 
         def format_error_msg(msg: str, lhs_types: List[Type], rhs_types: List[Type]) -> str:
             error_msg = ""

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4027,7 +4027,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
         subtype = get_proper_type(subtype)
         supertype = get_proper_type(supertype)
-
+        if self.check_tuple_assignment(subtype, supertype, context, msg,
+                                       subtype_label, supertype_label, code=code):
+            return False
         if self.should_suppress_optional_error([subtype]):
             return False
         extra_info = []  # type: List[str]
@@ -4377,6 +4379,22 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         yes_map, no_map = conditional_type_map(expr, vartype, type)
         yes_map, no_map = map(convert_to_typetype, (yes_map, no_map))
         return yes_map, no_map
+
+    def check_tuple_assignment(self,
+                               subtype: ProperType,
+                               supertype: ProperType,
+                               context: Context,
+                               msg: str = message_registry.INCOMPATIBLE_TYPES,
+                               subtype_label: Optional[str] = None,
+                               supertype_label: Optional[str] = None,
+                               code: Optional[ErrorCode] = None) -> bool:
+        if (isinstance(subtype, TupleType)
+            and isinstance(supertype, Instance)
+                and supertype.type.fullname == 'builtins.tuple'):
+            # TODO:
+            self.fail("error message", context, code=code)
+            return True
+        return False
 
 
 def conditional_type_map(expr: Expression,

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4027,7 +4027,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
         subtype = get_proper_type(subtype)
         supertype = get_proper_type(supertype)
-        if self.check_tuple_assignment(subtype, supertype, context, msg,
+        if self.check_long_tuple_assignment(subtype, supertype, context, msg,
                                        subtype_label, supertype_label, code=code):
             return False
         if self.should_suppress_optional_error([subtype]):
@@ -4388,32 +4388,49 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                                subtype_label: Optional[str] = None,
                                supertype_label: Optional[str] = None,
                                code: Optional[ErrorCode] = None) -> bool:
+
+        def format_error_msg(msg: str, lhs_types: List[Type], rhs_types: List[Type]) -> str:
+            error_msg = ""
+            error_cnt = 0
+            for i, (lhs_t, rhs_t) in enumerate(zip(lhs_types, rhs_types)):
+                if not is_subtype(lhs_t, rhs_t):
+                    if error_cnt < 3:
+                        error_msg += 'expression Tuple item {} has type "{}"; "{}" expected; '\
+                            .format(str(i), format_type_bare(rhs_t), format_type_bare(lhs_t))
+                    error_cnt += 1
+            error_msg += 'Total {} items are incompatible; {} items are omitted'.format(
+                        str(error_cnt), str(max(0, error_cnt - 3)))
+            return msg + ' (' + error_msg + ')'
+
+        def format_long_tuple_type(typ: TupleType) -> str:
+            item_cnt = len(typ.items)
+            if item_cnt > 10:
+                return 'Tuple[{}, {}, ... <{} more items>]'\
+                        .format(format_type_bare(typ.items[0]),
+                            format_type_bare(typ.items[1]), str(item_cnt - 2))
+            else:
+                return format_type_bare(typ)
+
         if isinstance(subtype, TupleType):
-            if isinstance(supertype, Instance) and supertype.type.fullname == 'builtins.tuple' and subtype:
+            if (len(subtype.items) > 10 and
+                isinstance(supertype, Instance) and
+                    supertype.type.fullname == 'builtins.tuple'):
                 lhs_type = supertype.args[0]
-                error_msg = ""
-                error_cnt = 0
-                for i, t in enumerate(subtype.items):
-                    if not is_subtype(t, lhs_type):
-                        if error_cnt < 3: 
-                            error_msg += 'Tuple item {} has type "{}"; "{}" expected; '.format(
-                                                    str(i), format_type_bare(t), format_type_bare(lhs_type))
-                        error_cnt += 1
-                error_msg += 'Total {} items are incompatible; {} items are omitted'.format(str(error_cnt), str(max(0, error_cnt - 3)))
-                self.fail(msg + ' (' + error_msg + ')', context, code=code)
+                lhs_types = [lhs_type] * len(subtype.items)
+                error_msg = format_error_msg(msg, lhs_types, subtype.items)
+                self.fail(error_msg, context, code=code)
                 return True
-            elif isinstance(supertype, TupleType):
-                lhs_types = supertype.items
-                rhs_types = subtype.items
-                error_msg = ""
-                error_cnt = 0
-                for i, l_t, r_t in enumerate(zip(lhs_types, rhs_types)):
-                    if not is_subtype(l_t, r_t):
-                        if error_cnt < 3:
-                            error_msg += 'Tuple item {} has type "{}"; "{}" expected; '.format(
-                                                    str(i), format_type_bare(t), format_type_bare(lhs_type))
-                        error_cnt += 1
-                self.fail("error message", context, code=code)
+            elif (isinstance(supertype, TupleType) and
+                    (len(subtype.items) > 10 or len(supertype.items) > 10)):
+                if len(subtype.items) != len(supertype.items):
+                    if supertype_label is not None and subtype_label is not None:
+                        error_msg = "{} ({} {}, {} {})".format(msg, subtype_label,
+                                        format_long_tuple_type(subtype), supertype_label,
+                                        format_long_tuple_type(supertype))
+                        self.fail(error_msg, context, code=code)
+                        return True
+                error_msg = format_error_msg(msg, supertype.items, subtype.items)
+                self.fail(error_msg, context, code=code)
                 return True
         return False
 

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4395,7 +4395,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                     supertype.type.fullname == 'builtins.tuple'):
                 lhs_type = supertype.args[0]
                 lhs_types = [lhs_type] * len(subtype.items)
-                self.generate_imcompatiable_tuple_error(lhs_types,
+                self.generate_incompatible_tuple_error(lhs_types,
                                     subtype.items, context, msg, code)
                 return True
             elif (isinstance(supertype, TupleType) and
@@ -4407,13 +4407,13 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                                         self.format_long_tuple_type(supertype))
                         self.fail(error_msg, context, code=code)
                         return True
-                self.generate_imcompatiable_tuple_error(supertype.items,
+                self.generate_incompatible_tuple_error(supertype.items,
                                     subtype.items, context, msg, code)
                 return True
         return False
 
     def format_long_tuple_type(self, typ: TupleType) -> str:
-        """Format very long tuple type using in ellipsis notation"""
+        """Format very long tuple type using an ellipsis notation"""
         item_cnt = len(typ.items)
         if item_cnt > 10:
             return 'Tuple[{}, {}, ... <{} more items>]'\
@@ -4422,22 +4422,20 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         else:
             return format_type_bare(typ)
 
-    def generate_imcompatiable_tuple_error(self,
-                                           lhs_types: List[Type],
-                                           rhs_types: List[Type],
-                                           context: Context,
-                                           msg: str = message_registry.INCOMPATIBLE_TYPES,
-                                           code: Optional[ErrorCode] = None) -> None:
-        """Generate error message for individual imcompatiable tuple pairs"""
+    def generate_incompatible_tuple_error(self,
+                                          lhs_types: List[Type],
+                                          rhs_types: List[Type],
+                                          context: Context,
+                                          msg: str = message_registry.INCOMPATIBLE_TYPES,
+                                          code: Optional[ErrorCode] = None) -> None:
+        """Generate error message for individual incompatible tuple pairs"""
         error_cnt = 0
         notes = []  # List[str]
-        notes_items = []  # List[Type]
         for i, (lhs_t, rhs_t) in enumerate(zip(lhs_types, rhs_types)):
             if not is_subtype(lhs_t, rhs_t):
                 if error_cnt < 3:
-                    notes.append('Expression Tuple item {} has type "{}"; "{}" expected; '
+                    notes.append('Expression tuple item {} has type "{}"; "{}" expected; '
                         .format(str(i), format_type_bare(rhs_t), format_type_bare(lhs_t)))
-                    notes_items.append(rhs_t)
                 error_cnt += 1
 
         error_msg = msg + ' ({} tuple items are incompatible'.format(str(error_cnt))
@@ -4446,7 +4444,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         else:
             error_msg += ')'
         self.fail(error_msg, context, code=code)
-        for note, item in zip(notes, notes_items):
+        for note in notes:
             self.note(note, context, code=code)
 
 

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4381,59 +4381,73 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         return yes_map, no_map
 
     def check_long_tuple_assignment(self,
-                               subtype: ProperType,
-                               supertype: ProperType,
-                               context: Context,
-                               msg: str = message_registry.INCOMPATIBLE_TYPES,
-                               subtype_label: Optional[str] = None,
-                               supertype_label: Optional[str] = None,
-                               code: Optional[ErrorCode] = None) -> bool:
+                                    subtype: ProperType,
+                                    supertype: ProperType,
+                                    context: Context,
+                                    msg: str = message_registry.INCOMPATIBLE_TYPES,
+                                    subtype_label: Optional[str] = None,
+                                    supertype_label: Optional[str] = None,
+                                    code: Optional[ErrorCode] = None) -> bool:
         """Check if assignment error using long tuple initializers."""
-
-        def format_error_msg(msg: str, lhs_types: List[Type], rhs_types: List[Type]) -> str:
-            error_msg = ""
-            error_cnt = 0
-            for i, (lhs_t, rhs_t) in enumerate(zip(lhs_types, rhs_types)):
-                if not is_subtype(lhs_t, rhs_t):
-                    if error_cnt < 3:
-                        error_msg += 'expression Tuple item {} has type "{}"; "{}" expected; '\
-                            .format(str(i), format_type_bare(rhs_t), format_type_bare(lhs_t))
-                    error_cnt += 1
-            error_msg += 'Total {} items are incompatible; {} items are omitted'.format(
-                        str(error_cnt), str(max(0, error_cnt - 3)))
-            return msg + ' (' + error_msg + ')'
-
-        def format_long_tuple_type(typ: TupleType) -> str:
-            item_cnt = len(typ.items)
-            if item_cnt > 10:
-                return 'Tuple[{}, {}, ... <{} more items>]'\
-                        .format(format_type_bare(typ.items[0]),
-                            format_type_bare(typ.items[1]), str(item_cnt - 2))
-            else:
-                return format_type_bare(typ)
-
         if isinstance(subtype, TupleType):
             if (len(subtype.items) > 10 and
                 isinstance(supertype, Instance) and
                     supertype.type.fullname == 'builtins.tuple'):
                 lhs_type = supertype.args[0]
                 lhs_types = [lhs_type] * len(subtype.items)
-                error_msg = format_error_msg(msg, lhs_types, subtype.items)
-                self.fail(error_msg, context, code=code)
+                self.generate_imcompatiable_tuple_error(lhs_types,
+                                    subtype.items, context, msg, code)
                 return True
             elif (isinstance(supertype, TupleType) and
                     (len(subtype.items) > 10 or len(supertype.items) > 10)):
                 if len(subtype.items) != len(supertype.items):
                     if supertype_label is not None and subtype_label is not None:
                         error_msg = "{} ({} {}, {} {})".format(msg, subtype_label,
-                                        format_long_tuple_type(subtype), supertype_label,
-                                        format_long_tuple_type(supertype))
+                                        self.format_long_tuple_type(subtype), supertype_label,
+                                        self.format_long_tuple_type(supertype))
                         self.fail(error_msg, context, code=code)
                         return True
-                error_msg = format_error_msg(msg, supertype.items, subtype.items)
-                self.fail(error_msg, context, code=code)
+                self.generate_imcompatiable_tuple_error(supertype.items,
+                                    subtype.items, context, msg, code)
                 return True
         return False
+
+    def format_long_tuple_type(self, typ: TupleType) -> str:
+        """Format very long tuple type using in ellipsis notation"""
+        item_cnt = len(typ.items)
+        if item_cnt > 10:
+            return 'Tuple[{}, {}, ... <{} more items>]'\
+                    .format(format_type_bare(typ.items[0]),
+                        format_type_bare(typ.items[1]), str(item_cnt - 2))
+        else:
+            return format_type_bare(typ)
+
+    def generate_imcompatiable_tuple_error(self,
+                                           lhs_types: List[Type],
+                                           rhs_types: List[Type],
+                                           context: Context,
+                                           msg: str = message_registry.INCOMPATIBLE_TYPES,
+                                           code: Optional[ErrorCode] = None) -> None:
+        """Generate error message for individual imcompatiable tuple pairs"""
+        error_cnt = 0
+        notes = []  # List[str]
+        notes_items = []  # List[Type]
+        for i, (lhs_t, rhs_t) in enumerate(zip(lhs_types, rhs_types)):
+            if not is_subtype(lhs_t, rhs_t):
+                if error_cnt < 3:
+                    notes.append('Expression Tuple item {} has type "{}"; "{}" expected; '
+                        .format(str(i), format_type_bare(rhs_t), format_type_bare(lhs_t)))
+                    notes_items.append(rhs_t)
+                error_cnt += 1
+
+        error_msg = msg + ' ({} tuple items are incompatible'.format(str(error_cnt))
+        if error_cnt - 3 > 0:
+            error_msg += '; {} items are omitted)'.format(str(error_cnt - 3))
+        else:
+            error_msg += ')'
+        self.fail(error_msg, context, code=code)
+        for note, item in zip(notes, notes_items):
+            self.note(note, context, code=code)
 
 
 def conditional_type_map(expr: Expression,

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4380,7 +4380,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         yes_map, no_map = map(convert_to_typetype, (yes_map, no_map))
         return yes_map, no_map
 
-    def check_tuple_assignment(self,
+    def check_long_tuple_assignment(self,
                                subtype: ProperType,
                                supertype: ProperType,
                                context: Context,
@@ -4388,12 +4388,33 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                                subtype_label: Optional[str] = None,
                                supertype_label: Optional[str] = None,
                                code: Optional[ErrorCode] = None) -> bool:
-        if (isinstance(subtype, TupleType)
-            and isinstance(supertype, Instance)
-                and supertype.type.fullname == 'builtins.tuple'):
-            # TODO:
-            self.fail("error message", context, code=code)
-            return True
+        if isinstance(subtype, TupleType):
+            if isinstance(supertype, Instance) and supertype.type.fullname == 'builtins.tuple' and subtype:
+                lhs_type = supertype.args[0]
+                error_msg = ""
+                error_cnt = 0
+                for i, t in enumerate(subtype.items):
+                    if not is_subtype(t, lhs_type):
+                        if error_cnt < 3: 
+                            error_msg += 'Tuple item {} has type "{}"; "{}" expected; '.format(
+                                                    str(i), format_type_bare(t), format_type_bare(lhs_type))
+                        error_cnt += 1
+                error_msg += 'Total {} items are incompatible; {} items are omitted'.format(str(error_cnt), str(max(0, error_cnt - 3)))
+                self.fail(msg + ' (' + error_msg + ')', context, code=code)
+                return True
+            elif isinstance(supertype, TupleType):
+                lhs_types = supertype.items
+                rhs_types = subtype.items
+                error_msg = ""
+                error_cnt = 0
+                for i, l_t, r_t in enumerate(zip(lhs_types, rhs_types)):
+                    if not is_subtype(l_t, r_t):
+                        if error_cnt < 3:
+                            error_msg += 'Tuple item {} has type "{}"; "{}" expected; '.format(
+                                                    str(i), format_type_bare(t), format_type_bare(lhs_type))
+                        error_cnt += 1
+                self.fail("error message", context, code=code)
+                return True
         return False
 
 

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -1387,6 +1387,77 @@ class MessageBuilder:
                       .format(len(conflicts) - max_items),
                       context, offset=offset, code=code)
 
+    def try_report_long_tuple_assignment_error(self,
+                                               subtype: ProperType,
+                                               supertype: ProperType,
+                                               context: Context,
+                                               msg: str = message_registry.INCOMPATIBLE_TYPES,
+                                               subtype_label: Optional[str] = None,
+                                               supertype_label: Optional[str] = None,
+                                               code: Optional[ErrorCode] = None) -> bool:
+        """Try to generate meaningful error message for very long tuple assignment
+
+        Returns a bool: True when generating long tuple assignment error,
+        False when no such error reported
+        """
+        if isinstance(subtype, TupleType):
+            if (len(subtype.items) > 10 and
+                isinstance(supertype, Instance) and
+                    supertype.type.fullname == 'builtins.tuple'):
+                lhs_type = supertype.args[0]
+                lhs_types = [lhs_type] * len(subtype.items)
+                self.generate_incompatible_tuple_error(lhs_types,
+                                    subtype.items, context, msg, code)
+                return True
+            elif (isinstance(supertype, TupleType) and
+                    (len(subtype.items) > 10 or len(supertype.items) > 10)):
+                if len(subtype.items) != len(supertype.items):
+                    if supertype_label is not None and subtype_label is not None:
+                        error_msg = "{} ({} {}, {} {})".format(msg, subtype_label,
+                                        self.format_long_tuple_type(subtype), supertype_label,
+                                        self.format_long_tuple_type(supertype))
+                        self.fail(error_msg, context, code=code)
+                        return True
+                self.generate_incompatible_tuple_error(supertype.items,
+                                    subtype.items, context, msg, code)
+                return True
+        return False
+
+    def format_long_tuple_type(self, typ: TupleType) -> str:
+        """Format very long tuple type using an ellipsis notation"""
+        item_cnt = len(typ.items)
+        if item_cnt > 10:
+            return 'Tuple[{}, {}, ... <{} more items>]'\
+                    .format(format_type_bare(typ.items[0]),
+                        format_type_bare(typ.items[1]), str(item_cnt - 2))
+        else:
+            return format_type_bare(typ)
+
+    def generate_incompatible_tuple_error(self,
+                                          lhs_types: List[Type],
+                                          rhs_types: List[Type],
+                                          context: Context,
+                                          msg: str = message_registry.INCOMPATIBLE_TYPES,
+                                          code: Optional[ErrorCode] = None) -> None:
+        """Generate error message for individual incompatible tuple pairs"""
+        error_cnt = 0
+        notes = []  # List[str]
+        for i, (lhs_t, rhs_t) in enumerate(zip(lhs_types, rhs_types)):
+            if not is_subtype(lhs_t, rhs_t):
+                if error_cnt < 3:
+                    notes.append('Expression tuple item {} has type "{}"; "{}" expected; '
+                        .format(str(i), format_type_bare(rhs_t), format_type_bare(lhs_t)))
+                error_cnt += 1
+
+        error_msg = msg + ' ({} tuple items are incompatible'.format(str(error_cnt))
+        if error_cnt - 3 > 0:
+            error_msg += '; {} items are omitted)'.format(str(error_cnt - 3))
+        else:
+            error_msg += ')'
+        self.fail(error_msg, context, code=code)
+        for note in notes:
+            self.note(note, context, code=code)
+
 
 def quote_type_string(type_string: str) -> str:
     """Quotes a type representation for use in messages."""

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1245,14 +1245,47 @@ reveal_type(a + b)  # N: Revealed type is 'Tuple[builtins.int, builtins.str, bui
 [case testAssigningWithLongTupleInitializer]
 from typing import Tuple
 
+# long initializer assignment with few mismatches
 t: Tuple[int, ...] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "9", "10", 11, "12", "13", "12", 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
-    30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
-    3, 31, 32, 33, 34, 35, 36, 37, 3, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
-    30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 4, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
-    30, 31, 32, 33, 34, 234234, 36, 37, 38, 39, 40, 1, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56,
-    57, 58, 59, 60, 61, 62, 63
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", 11, 12, 14, 15,
+    16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
+    31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
 )
+
+# long initializer assignment with more mismatches
+t1: Tuple[int, ...] = (
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", "str", "str", "str",
+    16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
+    31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
+)
+
+# short tuple initializer assignment
+t2: Tuple[int, ...] = (1, 2, "s", 4)
+
+# long initializer assignment with few mismatches, no ellipsis
+t3: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
+    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, "str", "str",
+)
+
+# long initializer assignment with more mismatches, no ellipsis
+t4: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str",
+)
+
+# short tuple initializer assignment, no ellipsis
+t5: Tuple[int, int] = (1, 2, "s", 4)
+
+# long initializer assignment with mismatched pairs
+t6: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", 1, 1, 1, 1, 1
+)
+
 [builtins fixtures/tuple.pyi]
 [out]
-main:4: error: Incompatible types in assignment (Tuple item 8 has type "str"; "int" expected; Tuple item 9 has type "str"; "int" expected; Tuple item 11 has type "str"; "int" expected; Total 5 items are incompatible; 2 items are omitted)
+main:5: error: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 3 items are incompatible; 0 items are omitted)
+main:12: error: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 7 items are incompatible; 4 items are omitted)
+main:18: error: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, ...]")
+main:22: error: Incompatible types in assignment (expression Tuple item 10 has type "str"; "int" expected; expression Tuple item 11 has type "str"; "int" expected; Total 2 items are incompatible; 0 items are omitted)
+main:27: error: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 4 items are incompatible; 1 items are omitted)
+main:31: error: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, int]")
+main:35: error: Incompatible types in assignment (expression has type Tuple[int, int, ... <15 more items>], variable has type Tuple[int, int, ... <10 more items>]) 

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1241,3 +1241,18 @@ b = ("bar", 7)
 reveal_type(a + b)  # N: Revealed type is 'Tuple[builtins.int, builtins.str, builtins.int, builtins.str, builtins.int]'
 
 [builtins fixtures/tuple.pyi]
+
+[case testAssigningWithLongTupleInitializer]
+from typing import Tuple
+
+t: Tuple[int, ...] = (
+    1, 2, 3, 4, 5, 6, 7, 8, "9", "10", 11, "12", "13", "12", 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
+    30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
+    3, 31, 32, 33, 34, 35, 36, 37, 3, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
+    30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 4, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
+    30, 31, 32, 33, 34, 234234, 36, 37, 38, 39, 40, 1, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56,
+    57, 58, 59, 60, 61, 62, 63
+)
+[builtins fixtures/tuple.pyi]
+[out]
+main:4: error: Incompatible types in assignment (Tuple item 8 has type "str"; "int" expected; Tuple item 9 has type "str"; "int" expected; Tuple item 11 has type "str"; "int" expected; Total 5 items are incompatible; 2 items are omitted)

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1247,45 +1247,37 @@ from typing import Tuple
 
 # long initializer assignment with few mismatches
 t: Tuple[int, ...] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", 11, 12, 14, 15,
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", 11, 12, 14, 15,  # E: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 3 items are incompatible; 0 items are omitted)
     16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
     31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
 )
 
 # long initializer assignment with more mismatches
 t1: Tuple[int, ...] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", "str", "str", "str",
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", "str", "str", "str",  # E: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 7 items are incompatible; 4 items are omitted)
     16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
     31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
 )
 
 # short tuple initializer assignment
-t2: Tuple[int, ...] = (1, 2, "s", 4)
+t2: Tuple[int, ...] = (1, 2, "s", 4)  # E: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, ...]")
 
 # long initializer assignment with few mismatches, no ellipsis
 t3: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
-    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, "str", "str",
+    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, "str", "str",  # E: Incompatible types in assignment (expression Tuple item 10 has type "str"; "int" expected; expression Tuple item 11 has type "str"; "int" expected; Total 2 items are incompatible; 0 items are omitted)
 )
 
 # long initializer assignment with more mismatches, no ellipsis
 t4: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str",
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str",  # E: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 4 items are incompatible; 1 items are omitted)
 )
 
 # short tuple initializer assignment, no ellipsis
-t5: Tuple[int, int] = (1, 2, "s", 4)
+t5: Tuple[int, int] = (1, 2, "s", 4)  # E: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, int]")
 
 # long initializer assignment with mismatched pairs
 t6: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", 1, 1, 1, 1, 1
+    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", 1, 1, 1, 1, 1  # E: Incompatible types in assignment (expression has type Tuple[int, int, ... <15 more items>], variable has type Tuple[int, int, ... <10 more items>]) 
 )
 
 [builtins fixtures/tuple.pyi]
-[out]
-main:5: error: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 3 items are incompatible; 0 items are omitted)
-main:12: error: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 7 items are incompatible; 4 items are omitted)
-main:18: error: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, ...]")
-main:22: error: Incompatible types in assignment (expression Tuple item 10 has type "str"; "int" expected; expression Tuple item 11 has type "str"; "int" expected; Total 2 items are incompatible; 0 items are omitted)
-main:27: error: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 4 items are incompatible; 1 items are omitted)
-main:31: error: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, int]")
-main:35: error: Incompatible types in assignment (expression has type Tuple[int, int, ... <15 more items>], variable has type Tuple[int, int, ... <10 more items>]) 

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1246,38 +1246,41 @@ reveal_type(a + b)  # N: Revealed type is 'Tuple[builtins.int, builtins.str, bui
 from typing import Tuple
 
 # long initializer assignment with few mismatches
-t: Tuple[int, ...] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", 11, 12, 14, 15,  # E: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 3 items are incompatible; 0 items are omitted)
-    16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
-    31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
-)
+t: Tuple[int, ...] = (1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", 11) \
+                                                                    # E: Incompatible types in assignment (3 tuple items are incompatible) \
+                                                                    # N: Expression Tuple item 8 has type "str"; "int" expected; \
+                                                                    # N: Expression Tuple item 9 has type "str"; "int" expected; \
+                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected;
 
 # long initializer assignment with more mismatches
-t1: Tuple[int, ...] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", "str", "str", "str",  # E: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 7 items are incompatible; 4 items are omitted)
-    16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
-    31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
-)
+t1: Tuple[int, ...] = (1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str") \
+                                                                    # E: Incompatible types in assignment (4 tuple items are incompatible; 1 items are omitted) \
+                                                                    # N: Expression Tuple item 8 has type "str"; "int" expected; \
+                                                                    # N: Expression Tuple item 9 has type "str"; "int" expected; \
+                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected;
 
 # short tuple initializer assignment
-t2: Tuple[int, ...] = (1, 2, "s", 4)  # E: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, ...]")
+t2: Tuple[int, ...] = (1, 2, "s", 4) \
+                                # E: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, ...]")
 
 # long initializer assignment with few mismatches, no ellipsis
-t3: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
-    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, "str", "str",  # E: Incompatible types in assignment (expression Tuple item 10 has type "str"; "int" expected; expression Tuple item 11 has type "str"; "int" expected; Total 2 items are incompatible; 0 items are omitted)
-)
+t3: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, "str", "str") \
+                                                                    # E: Incompatible types in assignment (2 tuple items are incompatible) \
+                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected; \
+                                                                    # N: Expression Tuple item 11 has type "str"; "int" expected;
 
 # long initializer assignment with more mismatches, no ellipsis
-t4: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str",  # E: Incompatible types in assignment (expression Tuple item 8 has type "str"; "int" expected; expression Tuple item 9 has type "str"; "int" expected; expression Tuple item 10 has type "str"; "int" expected; Total 4 items are incompatible; 1 items are omitted)
-)
+t4: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str") \
+                                                                    # E: Incompatible types in assignment (4 tuple items are incompatible; 1 items are omitted) \
+                                                                    # N: Expression Tuple item 8 has type "str"; "int" expected; \
+                                                                    # N: Expression Tuple item 9 has type "str"; "int" expected; \
+                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected;
 
 # short tuple initializer assignment, no ellipsis
 t5: Tuple[int, int] = (1, 2, "s", 4)  # E: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, int]")
 
 # long initializer assignment with mismatched pairs
-t6: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (
-    1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", 1, 1, 1, 1, 1  # E: Incompatible types in assignment (expression has type Tuple[int, int, ... <15 more items>], variable has type Tuple[int, int, ... <10 more items>]) 
-)
+t6: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str", 1, 1, 1, 1, 1) \
+                                                                    # E: Incompatible types in assignment (expression has type Tuple[int, int, ... <15 more items>], variable has type Tuple[int, int, ... <10 more items>]) 
 
 [builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1248,16 +1248,16 @@ from typing import Tuple
 # long initializer assignment with few mismatches
 t: Tuple[int, ...] = (1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", 11) \
                                                                     # E: Incompatible types in assignment (3 tuple items are incompatible) \
-                                                                    # N: Expression Tuple item 8 has type "str"; "int" expected; \
-                                                                    # N: Expression Tuple item 9 has type "str"; "int" expected; \
-                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected;
+                                                                    # N: Expression tuple item 8 has type "str"; "int" expected; \
+                                                                    # N: Expression tuple item 9 has type "str"; "int" expected; \
+                                                                    # N: Expression tuple item 10 has type "str"; "int" expected;
 
 # long initializer assignment with more mismatches
 t1: Tuple[int, ...] = (1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str") \
                                                                     # E: Incompatible types in assignment (4 tuple items are incompatible; 1 items are omitted) \
-                                                                    # N: Expression Tuple item 8 has type "str"; "int" expected; \
-                                                                    # N: Expression Tuple item 9 has type "str"; "int" expected; \
-                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected;
+                                                                    # N: Expression tuple item 8 has type "str"; "int" expected; \
+                                                                    # N: Expression tuple item 9 has type "str"; "int" expected; \
+                                                                    # N: Expression tuple item 10 has type "str"; "int" expected;
 
 # short tuple initializer assignment
 t2: Tuple[int, ...] = (1, 2, "s", 4) \
@@ -1266,15 +1266,15 @@ t2: Tuple[int, ...] = (1, 2, "s", 4) \
 # long initializer assignment with few mismatches, no ellipsis
 t3: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, "str", "str") \
                                                                     # E: Incompatible types in assignment (2 tuple items are incompatible) \
-                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected; \
-                                                                    # N: Expression Tuple item 11 has type "str"; "int" expected;
+                                                                    # N: Expression tuple item 10 has type "str"; "int" expected; \
+                                                                    # N: Expression tuple item 11 has type "str"; "int" expected;
 
 # long initializer assignment with more mismatches, no ellipsis
 t4: Tuple[int, int, int, int, int, int, int, int, int, int, int, int] = (1, 2, 3, 4, 5, 6, 7, 8, "str", "str", "str", "str") \
                                                                     # E: Incompatible types in assignment (4 tuple items are incompatible; 1 items are omitted) \
-                                                                    # N: Expression Tuple item 8 has type "str"; "int" expected; \
-                                                                    # N: Expression Tuple item 9 has type "str"; "int" expected; \
-                                                                    # N: Expression Tuple item 10 has type "str"; "int" expected;
+                                                                    # N: Expression tuple item 8 has type "str"; "int" expected; \
+                                                                    # N: Expression tuple item 9 has type "str"; "int" expected; \
+                                                                    # N: Expression tuple item 10 has type "str"; "int" expected;
 
 # short tuple initializer assignment, no ellipsis
 t5: Tuple[int, int] = (1, 2, "s", 4)  # E: Incompatible types in assignment (expression has type "Tuple[int, int, str, int]", variable has type "Tuple[int, int]")

